design-proposals/klip-53-pull-queries-on-streams.md

```diff
@@ -0,0 +1,87 @@
+# KLIP 53 - Pull Queries on Streams
+
+**Author**: John Roesler (@vvcephei) | 
+**Release Target**: 0.22 | 
+**Status**: _In Discussion_ | 
+**Discussion**: _link to the design discussion PR_
+
+**tl;dr:** People should be able to issue pull (ad-hoc, self-terminating) queries on streams,
+           just as they can with tables.
+           
+## Motivation and background
+
+ksqlDB allows the creation of both pull queries (which are issued ad-hoc and run to completion like
+a typical database query) and push queries (which are long-running subscriptions that continuously).
+ksqlDB also has a dual table/stream data model. You can issue push queries on tables and streams, 
+and you can issue pull queries on tables, but you cannot currently issue pull queries on tables.
+
+Previously, we believed it was unnecessary to implement pull queries on streams because for a
+stream, the "current state" is equivalent to the entire history of records. In other words, a
+pull query on a stream would return the same thing as a push query on a stream, if it is configured
+to start at the beginning of the stream and if it's cancelled once it reaches the end.
+
+In practice, though, it is not clear when you have reached the end of a stream, so you cannot
+easily determine when to cancel the query. Additionally, it is cumbersome to have to configure
+push queries to start from the beginning of the stream each time you want to issue one of these
+pull-type queries.
+
+## What is in scope
+
+* Queries should be supported on STREAM objects without “EMIT CHANGES” at the end
+* Pull queries should start at the beginning of the stream
+* Pull queries should terminate at the (current) end of the stream
+* Pull queries on streams will support the same range of query operations as they
+  do on tables. In other words, you will be able to filter the stream using a WHERE clause.
+
+## What is not in scope
+
+* Pull queries on streams will not currently support grouping/aggregations.
+  This would be useful but is left as future work.
+  Note that aggregations are also not currently supported on table pull queries.
```

Comment:
What about joins ?

Reply:
Joins are also not supported in pull queries. It's good to mention. Thanks!